crates/ty_python_semantic/src/types/signatures.rs

```diff
@@ -1404,6 +1332,14 @@ impl<'db> Signature<'db> {
             return ConstraintSet::from(true);
         }
 
+        // The top signature is supertype of (and assignable from) all other signatures. It is a
+        // subtype of no signature except itself, and assignable only to the gradual signature.
+        if other.parameters.is_top() {
+            return ConstraintSet::from(true);
+        } else if self.parameters.is_top() && !other.parameters.is_gradual() {
+            return ConstraintSet::from(false);
+        }
```

Comment:
no tests fail if I remove this, and I think the logic would be taken care of by the generalised fall-through below, so I think it's just a pure optimisation? But codspeed isn't reporting any speedup on this PR, so I'd be inclined to remove it, to keep the code simple (this method is already very complex)

```suggestion
        // The top signature is supertype of (and assignable from) all other signatures.
        if other.parameters.is_top() {
            return ConstraintSet::from(true);
        }
```

Reply:
This condition is necessary for correctness, though only in an edge case (and apparently one we're lacking test coverage of). Without it, we will wrongly consider a top callable to be a subtype of `(*object, **object) -> object` -- this is a side effect of how we give the top callable the parameters `(*object, **object)` in order to avoid redundant arity-checking errors when calling it.

I'll add a test case that fails without this.